docs/proposals/002-routing-processor-proposal/routing-processor.proto

```diff
@@ -0,0 +1,232 @@
+}
+
+// ===== Cache Info Types =====
+
+message CacheInfo {
+  bool cache_hit = 1;
+  float similarity_score = 2;
+  string cached_request_id = 3;
+  uint32 age = 4;
+}
+
+// ===== Mutation Types =====
+
+message HeaderValueOption {
+  Header header = 1;
+  bool append = 2;
+}
+
+message HeaderMutation {
+  repeated HeaderValueOption set_headers = 1;
+  repeated string remove_headers = 2;
+}
+
+message BodyMutation {
+  bytes body = 1;
+}
+
+// ===== Request/Response Types =====
+
+enum MessageType {
+  UNKNOWN = 0;
+  CLIENT_REQUEST = 1;
+  UPSTREAM_RESPONSE = 2;
+  CACHED_RESPONSE = 3;
+}
+
+message RequestMetadata {
+  string request_id = 1;
+  string model_name = 2;
+  float response_latency = 3;
+  uint64 token_count = 4;
+  float cache_score = 5;
+  optional int32 status_code = 6;
+}
+
+message RequestHeaders {
+  repeated Header headers = 1;
+  repeated Backend available_backends = 2;
+  string request_id = 3;
+}
+
+message RequestBody {
+  bytes body = 1;
+  bool end_of_stream = 2;
+  MessageType message_type = 3;
+  RequestMetadata metadata = 4;
+}
+
+message ProcessingRequest {
+  oneof request {
+    RequestHeaders headers = 1;
+    RequestBody body = 2;
+  }
+  ProcessingMode processing_mode = 3;
+}
+
+message CommonResponse {
+  HeaderMutation header_mutation = 1;
+  BodyMutation body_mutation = 2;
+  optional CacheInfo cache_info = 3;
+  optional string selected_backend = 4;
+}
+
+message ImmediateResponse {
+  int32 status_code = 1;
+  repeated Header headers = 2;
+  bytes body = 3;
+}
+
+message MessageProcessingResponse {
+  CommonResponse response = 1;
+}
+
+message ProcessingResponse {
+  oneof response {
+    ImmediateResponse immediate_response = 1;
+    MessageProcessingResponse message_processing = 2;
+  }
+  Metadata dynamic_metadata = 3;
+}
+
+// ===== Semantic Cache Types =====
+
+message SearchRequest {
+  repeated Message messages = 1;
+  float similarity_threshold = 2;
+  string model = 3;
+}
+
+message SearchResponse {
+  bool found = 1;
+  repeated Message response_messages = 2;
+  float similarity_score = 3;
+  Usage usage = 4;
+}
+
+message StoreChatRequest {
+  repeated Message request_messages = 1;
+  repeated Message response_messages = 2;
+  string model = 3;
+  string ttl = 4;
+  Usage usage = 5;
+}
+
+message StoreChatResponse {
+  bool success = 1;
+  string error = 2;
+}
+
+// ===== Capability Types =====
+
+message CapabilitiesRequest {}
+
+message CapabilitiesResponse {
+  bool semantic_cache_supported = 1;
+  bool model_selection_supported = 2;
+  bool immediate_response_supported = 3;
+}
+
+// ===== Services =====
+
+service RoutingProcessor {
+  rpc ExternalProcess(stream ProcessingRequest) returns (stream ProcessingResponse);
+  rpc GetCapabilities(CapabilitiesRequest) returns (CapabilitiesResponse);
+}
+
+service SemanticCacheService {
+  rpc SearchCache(SearchRequest) returns (SearchResponse) {}
+  rpc StoreChat(StoreChatRequest) returns (StoreChatResponse) {}
```

Comment:
In developing the PoC, I also found we can provide yet another service to first store the request message, gets a unique ID for the entry, and then later use that unique ID along with the response message to persist the complete request/response pair, reducing the traffic and the overhead of tracking request/response in extproc.